### PR TITLE
Add SCTP Network Policy tests

### PR DESCRIPTION
**Pull request by fedepaol**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:

This PR completes the checklist from [https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests), testing network policies with SCTP.

This is achieved by adding SCTP support to both the `agnhost` `connect` command and `porter` command, and to use then in the network policy tests to check if network policies are working.
This is complementary to https://github.com/kubernetes/kubernetes/pull/87926

In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

NONE

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

[KEP] : https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md

```

### CODE REVIEW

**Comment by fedepaol** [comment]
/cc @danwinship 
/cc @vishnoianil

I still need to complete the conversion to tables, but since it's pretty long I stopped waiting for feedbacks.

**Review by cmluciano — COMMENTED** [review]
/ok-to-test

**Comment by aojea** [comment]
> In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
> Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.

just my 2 cents, I had very bad experiences with Ginkgo trying to do things "out" of its way. I'd would try to aggregate the common code in functions and duplicate all the tests for each protocol 🤷‍♂

**Comment by fedepaol** [comment]
> > In order to maintain the ` [Feature:SCTPConnectivity]` and have the sctp tests optional, I started converting the current policy tests in a tabular form. To do that (and to maintain sctp optional), I removed the `BeforeEach` function and converted it to a custom one invoked with the different protocols in each table body.
> > Not sure if this is the right approach but I could not find a better one that prevents duplication while still allowing us to skip the SCTP tests.
> 
> just my 2 cents, I had very bad experiences with Ginkgo trying to do things "out" of its way. I'd would try to aggregate the common code in functions and duplicate all the tests for each protocol man_shrugging

Thanks for your suggestion, it was one of the possible alternatives I had in mind while writing this.
I'm fine with either of those, but I'll wait for other opinions so there is consensus before starting to refactor. The infrastructure is all there so it's "just" a matter of how I do rearrange the code.

**Comment by fedepaol** [comment]
@aojea I tried to rearrange under your suggestion, refactoring only the first test to avoid to do unnecessary work. Does this approach sound more reasonable?
I am afraid the pr won't be as easy to review as it was before code is getting moved around.

Others? @danwinship maybe?

File: test/images/agnhost/connect/connect.go
Comment (danwinship) [inline test/images/agnhost/connect/connect.go:106]: This doesn't seem to be correct; `MaxInitTimeout` → `sinit_max_init_timeo` is not a timeout on the connect operation as a whole, it's an upper bound on what the timeout can be for each attempt, of which there are 8 by default, starting at 3s each and then doubling until they hit the max. (eg, if you pass `--timeout 15` you'll get 3s + 6s + 12s + 15s + 15s + 15s + 15s + 15s = 1m36s).

There doesn't seem to be any good way to make the operation itself time out when you want it to without breaking the protocol so I think you're going to need to rewrite this to just do a timeout-less connect in a goroutine, with a separate timer.

File: test/images/agnhost/porter/porter.go
Comment (danwinship) [inline test/images/agnhost/porter/porter.go:132]: defer conn.Close()

**Review by danwinship — COMMENTED** [review]
reviewing just the agnhost parts, which I think you should split out to a separate PR now

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:84]: so as discussed on slack yesterday, I'd avoid major rewrites to the NetworkPolicy code. Just add one or two new SCTP tests and don't change the existing tests.

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:1862]: it seems like you ought to be able to share a lot more code between the tcp and sctp cases here

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:1880]: just use `agnhost connect` for both the tcp and sctp cases

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:1905]: `--timeout 1s` here is inconsistent with the TCP case which uses 8 seconds

File: test/images/agnhost/connect/connect.go
Comment (fedepaol) [reply test/images/agnhost/connect/connect.go:106]: Thanks for pointing out, I completely mis-interpreted it. I'll go with the goroutine approach.

**Comment by fedepaol** [comment]
Removed the agnhost section as it's now being done in https://github.com/kubernetes/kubernetes/pull/90797

Also, this PR will need the agnhost image to be promoted before it can land.

**Comment by fedepaol** [comment]
/hold cancel

Tested on a local OCP cluster with ovn-kubernetes as sdn:
[https://pastebin.com/jiqkvW4N](https://pastebin.com/jiqkvW4N)

**Comment by aojea** [comment]
it LGTM
is the agnost image ` "2.20"` already available @fedepaol ?

**Comment by fedepaol** [comment]
> it LGTM
> is the agnost image ` "2.20"` already available @fedepaol ?

Actually 2.21. I tested with it. And thanks to your comment I realized I forgot to force push the last change :P
Promoted in https://github.com/kubernetes/k8s.io/pull/983

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:2092]: change `ports` to `tcpPorts`

or have a single `ports` argument that takes a struct of port and protocol

File: test/e2e/network/network_policy.go
Comment (danwinship) [inline test/e2e/network/network_policy.go:1821]: This creates a server that's listening on both TCP and SCTP but you never use the TCP ports

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go]: Yes, I can't remember the original version (it was some time ago) but probably it's a leftover of a more complex approach to make all the existing tests viable for both protocols.

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go]: Brilliant! Thanks for the suggestion

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go]: Changed. I'd prefer to keep the variadic function so `sctpPort`s are optional and do not impact existing tests.

**Comment by fedepaol** [comment]
@danwinship changed, PTAL

**Comment by fedepaol** [comment]
Validation here https://pastebin.com/ta1cyKkv

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go:53]: I have it to have a type safe parameter to pass where I need to choose between tcp and sctp.
In `func testCanConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol connProtocol) {` for example. I can replace with a bare string if you prefer.

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go:1811]: Np, agree it's cleaner that way. I'll change it.

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go:2142]: It won't be working without the timeout as the sctp implementation uses a select.

**Comment by fedepaol** [comment]
Ok, I think things are converging. @danwinship please take another look.
I managed to put some more code together by using the same readiness probe for both protocols.

File: test/e2e/network/network_policy.go
Comment (danwinship) [reply test/e2e/network/network_policy.go]: oh... actually, that's a bug in the agnhost code; if you don't specify a timeout, it treats that as "time out immediately", but it should be "time out never" like in the TCP case

Comment (fedepaol) [reply test/e2e/network/network_policy.go]: ok, let me fix it then

Comment (fedepaol) [reply test/e2e/network/network_policy.go]: opened https://github.com/kubernetes/kubernetes/pull/92657

**Comment by fedepaol** [comment]
/hold
Depends on https://github.com/kubernetes/kubernetes/pull/92657

**Comment by danwinship** [comment]
> /hold
> Depends on #92657

That might take a while. I'm fine merging this now if you add back the theoretically-unnecessary `--timeout` flag

**Comment by fedepaol** [comment]
> > /hold
> > Depends on #92657
> 
> That might take a while. I'm fine merging this now if you add back the theoretically-unnecessary `--timeout` flag

Done! Sorry I missed this last comment, it was buried in CI reports and I was monitoring the other PR.

**Comment by fedepaol** [comment]
For getting approved the image bump up:
/cc @luxas @mkumatag @listx 
as they are listed as approvers in `test/utils/image/OWNERS` 

/cc @BenTheElder as he bumped the previous agnhost version.

**Comment by fedepaol** [comment]
> For getting approved the image bump up:
> /cc @luxas @mkumatag @listx
> as they are listed as approvers in `test/utils/image/OWNERS`
> 
> /cc @BenTheElder as he bumped the previous agnhost version.

@luxas @mkumatag @listx ping!

File: test/e2e/network/network_policy.go
Comment (spiffxp) [inline test/e2e/network/network_policy.go:2210]: Comments that would help future test maintainers: What's the rationale behind the magic numbers here?
- why retry 5 times and not more/less
- why timeout at 8s instead of more/less

File: test/e2e/network/network_policy.go
Comment (spiffxp) [inline test/e2e/network/network_policy.go:1794]: According to https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests these should be marked as `[Disruptive]`?

Comment (danwinship) [reply test/e2e/network/network_policy.go:1794]: Yes. And we should belatedly fix the other SCTPConnectivity tests to have that now too

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go:2210]: Those magic numbers were already there, I left them while changing the way we open the socket in a way that supports both TCP and SCTP.

I tried to look back up to https://github.com/kubernetes/kubernetes/commit/3471daf7599604561825c7133aee6cc76408d199 but could not find a proper reasoning behind that.

File: test/e2e/network/network_policy.go
Comment (fedepaol) [reply test/e2e/network/network_policy.go]: Done, and added a separate commit to mark the old ones disruptive as well.

File: test/e2e/network/network_policy.go
Comment (spiffxp) [reply test/e2e/network/network_policy.go:2232]: Fair enough, thanks for looking

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
/approve cancel
actually... wait

https://github.com/kubernetes/kubernetes/pull/94152 landed which bumped agnhost to 2.22, I am concerned this will move it back to 2.21

File: build/dependencies.yaml
Comment (spiffxp) [inline build/dependencies.yaml:13]: I believe if you rebase you'll find this is now 2.22

File: test/utils/image/manifest.go
Comment (spiffxp) [inline test/utils/image/manifest.go:203]:
```suggestion
	configs[Agnhost] = Config{promoterE2eRegistry, "agnhost", "2.22"}
```

File: build/dependencies.yaml
Comment (spiffxp) [reply build/dependencies.yaml:13]: 🤦 you did rebase, and I missed the context above, nevermind, this is fine

File: test/utils/image/manifest.go
Comment (spiffxp) [reply test/utils/image/manifest.go:203]: nevermind, this isn't strictly speaking necessary

**Comment by spiffxp** [comment]
/approve
ok, don't mind me, this is fine

**Comment by fedepaol** [comment]
> /approve
> ok, don't mind me, this is fine

:slightly_smiling_face:

**Comment by spiffxp** [comment]
/milestone v1.20
This is a test-only change; the tests introduced are gated behind Disruptive.

The main that concerns me from a will-this-be-disruptive perspective is the bump to agnhost:2.21.  But we're going to have to bump it eventually, so may as well do it now.  Also it looks like that bump would only pickup https://github.com/kubernetes/kubernetes/pull/90797 which looks unobjectionable to me

**Pull request merged**
